hefquin-engine/src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/FederationAccessManagerWithPersistedDiskCache.java

```diff
@@ -0,0 +1,374 @@
+package se.liu.ida.hefquin.engine.federation.access.impl;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+
+import se.liu.ida.hefquin.base.datastructures.Cache;
+import se.liu.ida.hefquin.base.datastructures.impl.cache.CacheEntry;
+import se.liu.ida.hefquin.base.datastructures.impl.cache.CacheEntryBase;
+import se.liu.ida.hefquin.base.datastructures.impl.cache.CacheEntryBaseFactory;
+import se.liu.ida.hefquin.base.datastructures.impl.cache.CacheEntryFactory;
+import se.liu.ida.hefquin.base.datastructures.impl.cache.CacheInvalidationPolicy;
+import se.liu.ida.hefquin.base.datastructures.impl.cache.CacheInvalidationPolicyAlwaysValid;
+import se.liu.ida.hefquin.base.datastructures.impl.cache.CachePolicies;
+import se.liu.ida.hefquin.base.datastructures.impl.cache.CacheReplacementPolicy;
+import se.liu.ida.hefquin.base.datastructures.impl.cache.CacheReplacementPolicyFactory;
+import se.liu.ida.hefquin.base.datastructures.impl.cache.CacheReplacementPolicyLRU;
+import se.liu.ida.hefquin.base.datastructures.impl.cache.GenericCacheImpl;
+import se.liu.ida.hefquin.base.query.TriplePattern;
+import se.liu.ida.hefquin.base.utils.Pair;
+import se.liu.ida.hefquin.engine.federation.BRTPFServer;
+import se.liu.ida.hefquin.engine.federation.FederationMember;
+import se.liu.ida.hefquin.engine.federation.Neo4jServer;
+import se.liu.ida.hefquin.engine.federation.SPARQLEndpoint;
+import se.liu.ida.hefquin.engine.federation.TPFServer;
+import se.liu.ida.hefquin.engine.federation.access.*;
+
+/**
+ * A FederationAccessManager implementation that incorporates persistent disk
+ * caching of SPARQL cardinality requests.
+ * 
+ * TODO: The implementation uses a simple serialization/deserialization
+ * strategy, where the file is stored to disk by writing the full map to disk
+ * for each update. This approach is not optimized for this task but is simply
+ * intended as a proof of concept. A future implementation should support
+ * standard cache configuration policys, such as time-based eviction (time to
+ * live), but should leverage an optimized persistence strategy, preferably
+ * leveraging a libary.
+ * 
+ * Note: Most of the classes/interfaces involved (e.g., DataRetrievalResponse,
+ * CardinalityResponse etc.) do not support serialization.
+ */
+public class FederationAccessManagerWithPersistedDiskCache implements FederationAccessManager
```

Comment:
It would be better to make this class a subclass of `FederationAccessManagerWithCache`. This way, you don't need to duplicate all the code of the `issueRequest` functions and of `getStats()` etc.

Reply:
Good idea, will do!

Reply:
[e70c5fa](https://github.com/LiUSemWeb/HeFQUIN/commit/e70c5fa74732d16912b290b94076b00320a005d6)